### DISCUSSION TITLE
Асинхронный DataService

### DISCUSSION THREAD

**Discussion by bratchikov** [Flexberry/NewPlatform.Flexberry.ORM#205] [Ideas]
Давайте сформулируем требования к реализации асинхронности в сервисе данных. Какие интерфейсы нужны, особенности реализации и т.п.

**Comment by archaim** [comment]
### Вопросы на подумать
_(этот пост будет обновляться, пишите мне если есть что добавить)_
1. Нужен ли объединённый интерфейс, который включает и синхронные и асинхронные методы.
2. Как насчёт поднять некоторые элементы SQLDataService на уровень базового интерфейса? А то он слишком его расширяет. Возможно, вместо IAsyncDataService нам вообще нужен IAsyncSQLDataService.
3. Подумать над интерфейсом методов вроде UpdateObject. Ref-параметры в асинхронных методах недоступны, изменение объектов в параметрах — не факт, что хорошая идея. А если возвращать изменённые копии в качестве результата, то получается неочевидное поведение, имхо.

**Comment by archaim** [comment]
В чате рассмотрели 4 варианта:

    Расширить IDataService новыми методами
    Расширить IDataService через наследование IAsyncDataService
    Реализовать отдельный интерфейс IAsyncDataService только с методами для асинхронной работы.
    Реализовать отдельный интерфейс IAsyncDataService с асинхронными и обновлёнными синхронными методами.

Большинство склоняется к 3, но Коля предлагает 2 для упрощения миграции. С главным все согласны — нужен отдельный интефрейс IAsyncDataService, который включает только методы для асинхронной работы.

**Comment by archaim** [comment]
Draft pull request: #212
